src/invidious/routes/api/manifest.cr

```diff
@@ -70,17 +70,22 @@ module Invidious::Routes::API::Manifest
               # OTF streams aren't supported yet (See https://github.com/TeamNewPipe/NewPipe/issues/2415)
               next if !(fmt.has_key?("indexRange") && fmt.has_key?("initRange"))
 
+              audio_track = fmt["audioTrack"]?.try &.as_h? || {} of String => JSON::Any
+              lang = audio_track["id"]?.try &.as_s.split('.')[0] || "und"
+              is_default = audio_track.has_key?("audioIsDefault") ? audio_track["audioIsDefault"].as_bool : i == 0
+              displayname = audio_track["displayName"]?.try &.as_s || "Unknown"
+
               # Different representations of the same audio should be groupped into one AdaptationSet.
               # However, most players don't support auto quality switching, so we have to trick them
               # into providing a quality selector.
               # See https://github.com/iv-org/invidious/issues/3074 for more details.
-              xml.element("AdaptationSet", id: i, mimeType: mime_type, startWithSAP: 1, subsegmentAlignment: true, label: fmt["bitrate"].to_s + "k") do
+              xml.element("AdaptationSet", id: i, mimeType: mime_type, startWithSAP: 1, subsegmentAlignment: true, label: displayname, lang: lang) do
```

Comment:
In the case of different representations of the same language you'll end up with two `AdaptationSet` with the same label and lang. I think that might confuse some downstream clients (and users) if two audio tracks end up having the same label and language.

For example see the dash manifest for this video with this PR https://youtube.comwatch?v=urevinis_PU

Reply:
Hello.

If I am not misunderstanding your statement, yes that could happen if audio streams weren't being filtered for mime type "audio/mp4".

However, since they are being filtered this way, I have only seen one unique language representation per this mime type from YT.

With this PR and the current state of inv and YT, I have not experienced the issue you describe. And in the case of your example, no audio stream selection options are provided (by video.js and in my case) because there is no audio track information from YT, so the lang attribute is set to "und" and the client falls back to the default behavior of auto-selecting the first stream. When audio track info is available, video.js provides only one audio option per language in all of my testing.

Thank you for your time and attention.

Reply:
This isn't just about video.js. There are downstream clients that use other media players which also rely on the dash manifests generated by Invidious.

Reply:
I'm also concerned that this could reintroduce issue #3074.

Although the first `AdaptationSet` should always have the highest bitrate, YouTube's structure is very finicky and subject to change. Even if clients will always select the first stream it doesn't mean that it'll always have the highest bitrate.

PR #3162 also seemed to introduce a quality selector for audio but it seems to have broke somewhere down the line. If it was ever fixed, this issue will present a UX problem even in Invidious and video.js with the labels being simply "Unknown"

Reply:
What's the way forward with this? What is the API that needs to be maintained for downstream clients? Does it just need some kind of unique ID?

There is also, of course, the possibility of doing a breaking change. We could argue that if clients can't support this any more then they need updating. Invidious is becoming more and more useless and more and more videos come with some AI-generated dub track.